In [1]:
#!pip install tensorflow==2.0.0-rc1
#!pip install tensorflow-gpu==2.0.0-rc1

In [2]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

## Load the data and do the preprocessing:

In [3]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

Next, one-hot code your target variable using the `to_categorical()` function from the `keras.utils` module:

In [4]:
from tensorflow.keras.utils import to_categorical

Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

Now, check the size of the dataset:

In [5]:
X_train[0].shape

(784,)

## 1) In this task, you'll implement several ANN models with different activation functions. Specifically, do the following:

## Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [6]:
# Create the model
tanh_model = Sequential()

tanh_model.add(Dense(128, input_shape=(784,), activation="tanh"))
tanh_model.add(Dense(64, activation="tanh"))
tanh_model.add(Dense(10, activation="softmax"))

In [7]:
tanh_model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [8]:
# Compole the model
tanh_model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [9]:
# Fit the model
tanh_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 23us/sample - loss: 1.0154 - accuracy: 0.7561
Epoch 2/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.5131 - accuracy: 0.8722
Epoch 3/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.4162 - accuracy: 0.8900
Epoch 4/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.3699 - accuracy: 0.8988
Epoch 5/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.3413 - accuracy: 0.9048
Epoch 6/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.3211 - accuracy: 0.9093
Epoch 7/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.3054 - accuracy: 0.9133
Epoch 8/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.2926 - accuracy: 0.9167
Epoch 9/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.2817 - accuracy: 0.9191
Epoch 10/20
60

In [10]:
# Print the model results
score = tanh_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.20621993271410466
Test accuracy: 0.9414


## Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [11]:
# Create the model
sig_model = Sequential()

sig_model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
sig_model.add(Dense(64, activation="sigmoid"))
sig_model.add(Dense(10, activation="softmax"))

In [12]:
sig_model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_4 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_5 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [13]:
# Compole the model
sig_model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [14]:
# Fit the model
sig_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 20us/sample - loss: 2.2931 - accuracy: 0.1413
Epoch 2/20
60000/60000 [==============================] - 1s 16us/sample - loss: 2.2274 - accuracy: 0.3208
Epoch 3/20
60000/60000 [==============================] - 1s 16us/sample - loss: 2.1593 - accuracy: 0.4775
Epoch 4/20
60000/60000 [==============================] - 1s 16us/sample - loss: 2.0640 - accuracy: 0.5705
Epoch 5/20
60000/60000 [==============================] - 1s 16us/sample - loss: 1.9315 - accuracy: 0.6026
Epoch 6/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.7628 - accuracy: 0.6310
Epoch 7/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.5752 - accuracy: 0.6619
Epoch 8/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.3937 - accuracy: 0.6897
Epoch 9/20
60000/60000 [==============================] - 1s 15us/sample - loss: 1.2361 - accuracy: 0.7218
Epoch 10/20
60

In [15]:
# Print the model results
score = sig_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.5442446291446685
Test accuracy: 0.8672


## Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [16]:
# Create the model
relu_model = Sequential()

relu_model.add(Dense(128, input_shape=(784,), activation="relu"))
relu_model.add(Dense(64, activation="relu"))
relu_model.add(Dense(10, activation="softmax"))

In [17]:
relu_model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_7 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_8 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [18]:
# Compole the model
relu_model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [19]:
# Fit the model
relu_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 21us/sample - loss: 1.2649 - accuracy: 0.6892
Epoch 2/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.4990 - accuracy: 0.8674
Epoch 3/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.3897 - accuracy: 0.8922
Epoch 4/20
60000/60000 [==============================] - 1s 15us/sample - loss: 0.3428 - accuracy: 0.9034
Epoch 5/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.3140 - accuracy: 0.9114
Epoch 6/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.2926 - accuracy: 0.9173
Epoch 7/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.2757 - accuracy: 0.9224
Epoch 8/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.2613 - accuracy: 0.9265
Epoch 9/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.2491 - accuracy: 0.9296
Epoch 10/20
60

In [20]:
# Print the model results
score = relu_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.1644923143029213
Test accuracy: 0.9514


## Compare the results of each model. Which activation function performed best?

* tanh = 94%
* sigmoid = 86%
* reLU = 95%

The reLU model performed the best

## 2) In this task, you'll implement the ANN models specified below. For each, use the hinge loss function as the loss function.

## Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the tanh activation function for each layer.

In [21]:
# Create the model
tanh_model = Sequential()

tanh_model.add(Dense(128, input_shape=(784,), activation="tanh"))
tanh_model.add(Dense(64, activation="tanh"))
tanh_model.add(Dense(10, activation="softmax"))

In [22]:
tanh_model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_10 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_11 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [23]:
# Compole the model
tanh_model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

In [24]:
# Fit the model
tanh_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
60000/60000 [==============================] - 1s 23us/sample - loss: 0.9863 - accuracy: 0.4142
Epoch 2/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.7743 - accuracy: 0.7005
Epoch 3/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.5852 - accuracy: 0.7995
Epoch 4/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.4630 - accuracy: 0.8465
Epoch 5/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3885 - accuracy: 0.8676
Epoch 6/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3425 - accuracy: 0.8777
Epoch 7/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3120 - accuracy: 0.8835
Epoch 8/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2901 - accuracy: 0.8888
Epoch 9/20
60000/60000 [===

In [25]:
# Print the model results
score = tanh_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.1893198136627674
Test accuracy: 0.9207


## Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the sigmoid activation function for each layer.

In [26]:
# Create the model
sig_model = Sequential()

sig_model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
sig_model.add(Dense(64, activation="sigmoid"))
sig_model.add(Dense(10, activation="softmax"))

In [27]:
sig_model.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_13 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_14 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [28]:
# Compole the model
sig_model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

In [29]:
# Fit the model
sig_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 22us/sample - loss: 1.0101 - accuracy: 0.0932
Epoch 2/20
60000/60000 [==============================] - 1s 17us/sample - loss: 1.0036 - accuracy: 0.1152
Epoch 3/20
60000/60000 [==============================] - 1s 16us/sample - loss: 1.0029 - accuracy: 0.1576
Epoch 4/20
60000/60000 [==============================] - 1s 16us/sample - loss: 1.0024 - accuracy: 0.2051
Epoch 5/20
60000/60000 [==============================] - 1s 18us/sample - loss: 1.0019 - accuracy: 0.2569
Epoch 6/20
60000/60000 [==============================] - 1s 17us/sample - loss: 1.0014 - accuracy: 0.3174
Epoch 7/20
60000/60000 [==============================] - 1s 16us/sample - loss: 1.0010 - accuracy: 0.3773
Epoch 8/20
60000/60000 [==============================] - 1s 16us/sample - loss: 1.0007 - accuracy: 0.4237
Epoch 9/20
60000/60000 [==============================] - 1s 16us/sample - loss: 1.0004 - accuracy: 0.4629
Epoch 10/20
60

In [30]:
# Print the model results
score = sig_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.997283199596405
Test accuracy: 0.6644


## Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the ReLU activation function for each layer.

In [31]:
# Create the model
relu_model = Sequential()

relu_model.add(Dense(128, input_shape=(784,), activation="relu"))
relu_model.add(Dense(64, activation="relu"))
relu_model.add(Dense(10, activation="softmax"))

In [32]:
relu_model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_16 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_17 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [33]:
# Compole the model
relu_model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

In [34]:
# Fit the model
relu_model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 24us/sample - loss: 1.0115 - accuracy: 0.3374
Epoch 2/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.9746 - accuracy: 0.5459
Epoch 3/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.8689 - accuracy: 0.5833
Epoch 4/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.6937 - accuracy: 0.7100
Epoch 5/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.5272 - accuracy: 0.8062
Epoch 6/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.4122 - accuracy: 0.8554
Epoch 7/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.3387 - accuracy: 0.8759
Epoch 8/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2949 - accuracy: 0.8867
Epoch 9/20
60000/60000 [==============================] - 1s 16us/sample - loss: 0.2672 - accuracy: 0.8932
Epoch 10/20
60

In [35]:
# Print the model results
score = relu_model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16599102637767793
Test accuracy: 0.9272


## Compare the results of each model with the result of the same model from the previous task. Which loss function performed best?

With crossentropy loss:
* tanh = 94%
* sigmoid = 86%
* reLU = 95%

With hinge loss:
* tanh = 92%
* sigmoid = 66%
* reLU = 92%

### All 3 activation types performed worse using hinge loss
### tanh and reLU performed the best using either crossentropy or hinge loss